### PR TITLE
Improved numerical accuracy for log_sum_exp, softmax, and log_softmax

### PR DESCRIPTION
**Pull request by helske**
## Summary

This PR concerns #2802, I already started to work on this on Stancon hackathon with @SteveBronder. 

The current primitives of `softmax`, `log_softmax` and `log_sum_exp` are numerically stable according to [this paper](https://academic.oup.com/imajna/article/41/4/2311/5893596). However, the `fwd` and `rev` specializations are numerically less stable, and their handling of non-finite values is not consistent with the primitives.

The new `fwd` and `rev` specializations of these functions rely on the stable `softmax` and `log_sum_exp` primitives rather than independently reimplementing them. To make this possible, I relaxed the `require_*` constraints on the `softmax` and `log_softmax` primitives so that they support also also matrix input (operating over all elements of the matrix). If this is not a good idea, alternatively `fwd` and `rev` could use reshaping/flattening or reimplement the stable softmax (see also below).

There is still potential room for improvement: 

1.  `fwd::log_softmax`,  `fwd::log_sum_exp`, and `rev::log_sum_exp` now compute both `softmax(x)` and `log_sum_exp(x)` which have overlap in computations which could be avoided by computing both in a single pass. 
2.  While the paper linked above concludes that the current `max + std::log((v_ref.array() - max).exp().sum());` is safe, their suggested algorithm 4.1 excludes the maximum from the sum and uses `log1p`, which, based on some standalone tests I made could make a difference in some cases. Basically we could do something like
```
const auto v_flat = v_ref.reshaped(); 
Eigen::Index k; 
const double max = v_flat.maxCoeff(&k); 
if (!std::isfinite(max)) { 
 return max; 
}
Eigen::ArrayXd w = (v_flat.array() - max).exp();
w(k) = 0.0; 
return max + std::log1p(w.sum());
```

## Tests

The new tests covers numerical stability and non-finite-value handling in the `fwd` and `rev` specializations of `log_softmax` and `log_sum_exp`. The tests cover inputs with large dynamic ranges, many small probabilities, and infinite values. These tests fail with the current implementations and pass with the proposed changes. At least on my setup, few of the tests have tolerances of 1e-12 which I'm not sure are actually portable. For writing the tests, I used help from LLM.

## Side Effects

I guess this now exposes support for matrices in `softmax` and `log_softmax`?

## Release notes

Improved numerical accuracy of `log_sum_exp`, `softmax` and `log_softmax` functions.

## Checklist

- [x] Copyright holder: Jouni Helske

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by WardBrian** [comment]
Hi @helske -- it looks like the actual changes to the source may be missing from your branch, only the test changes seem to have been checked in

**Comment by helske** [comment]
Yes I run out of time earlier today. Here's the actual changes.

Note: I did not change the current two-scalar versions of `fwd::log_sum_exp(x1, x2)` and `rev::log_sum_exp(x1, x2)`, but I think there is still a problem here as well, or at least inconsistency with the vector valued version:

```
template <typename T>
inline fvar<T> log_sum_exp(const fvar<T>& x1, const fvar<T>& x2) {
  return fvar<T>(log_sum_exp(x1.val_, x2.val_),
                 x1.d_ * inv_logit(-(x2.val_ - x1.val_))
                     + x2.d_ * inv_logit(-(x1.val_ - x2.val_)));
}
```
If `x1` is infinite and `x2` is finite, this returns infinite value but potentially finite derivative, whereas `log_sum_exp(T&& x)` version always returns `NaN`? But I'm not sure if this matters in practice at all given nonfinite value in both cases?

File: stan/math/fwd/fun/log_sum_exp.hpp [hunk 1]
```diff
@@ -56,11 +57,10 @@ inline auto log_sum_exp(T&& x) {
...
         using mat_type = Eigen::Matrix<T_fvar_inner, -1, -1>;
         mat_type vals = v.val();
-        mat_type exp_vals = vals.array().exp();
 
-        return fvar<T_fvar_inner>(
-            log_sum_exp(vals),
-            v.d().cwiseProduct(exp_vals).sum() / exp_vals.sum());
+        const auto probs = softmax(vals);
+        return fvar<T_fvar_inner>(log_sum_exp(vals),
+                                  v.d().cwiseProduct(probs).sum());
```
Comment (andrjohns) [inline stan/math/fwd/fun/log_sum_exp.hpp:63]:
```suggestion
        return fvar<T_fvar_inner>(log_sum_exp(vals),
                                  v.d().cwiseProduct(softmax(vals)).sum());
```

File: stan/math/fwd/fun/log_softmax.hpp [hunk 2]
```diff
@@ -28,27 +29,31 @@ inline auto log_softmax(T&& x) {
...
+ * input is empty
  */
-template <typename Vec, require_eigen_vector_vt<is_fvar, Vec>* = nullptr>
-inline auto log_softmax(Vec&& x) {
-  using vec = std::decay_t<Vec>;
-  constexpr int Rows = vec::RowsAtCompileTime;
-  constexpr int Cols = vec::ColsAtCompileTime;
-  using T = typename value_type_t<vec>::Scalar;
-  decltype(auto) x_ref = to_ref(std::forward<Vec>(x));
+template <typename Mat, require_eigen_vt<is_fvar, Mat>* = nullptr>
```
Comment (andrjohns) [inline stan/math/fwd/fun/log_softmax.hpp:39]: A couple of 'gotchas' to be aware of. When calling `.val()`/`value_of()`/`.d()` on a matrix of `fvar` types, the resulting expression is essentially a view over non-contiguous memory. This makes Eigen expressions less efficient and vectorisable, so if you need to use them multiple times then it's generally more performant to pay the copy-cost once so that the following expressions will be faster.

The size of expressions can be checked without triggering an evaluation (correct me if I'm wrong there @SteveBronder) so you can use `if (x.size() == 0)` for the early return, since `to_ref()` can evaluate expressions.

There are also a few helper functions in the math library for simplifying the construction which you could use. Putting it all together, a simplified implementation could look like:

```cpp
template <typename Mat, require_eigen_vt<is_fvar, Mat>* = nullptr>
inline plain_type_t<Mat> log_softmax(Mat&& x) {
  if (x.size() == 0) {
    return {};
  }
  decltype(auto) x_ref = to_ref(std::forward<Mat>(x));
  const auto x_val = value_of(x_ref).eval();
  const auto d_in = x_ref.d().eval();
  return to_fvar(x_val.array() - log_sum_exp(x_val),
                 d_in.array() - softmax(x_val).cwiseProduct(d_in).sum());
}
```

File: stan/math/fwd/fun/softmax.hpp [hunk 3]
```diff
@@ -29,24 +29,25 @@ inline auto softmax(T&& x) {
...
+ *   empty
  */
-template <typename Vec, require_eigen_vector_vt<is_fvar, Vec>* = nullptr>
-inline auto softmax(Vec&& x) {
-  using vec = std::decay_t<Vec>;
-  constexpr int Rows = vec::RowsAtCompileTime;
-  constexpr int Cols = vec::ColsAtCompileTime;
-  using T = typename value_type_t<vec>::Scalar;
-  decltype(auto) x_ref = to_ref(std::forward<Vec>(x));
+template <typename Mat, require_eigen_vt<is_fvar, Mat>* = nullptr>
```
Comment (andrjohns) [inline stan/math/fwd/fun/softmax.hpp:37]: Using the same approach from `log_softmax`, this could simplify down to:

```cpp
template <typename Mat, require_eigen_vt<is_fvar, Mat>* = nullptr>
inline plain_type_t<Mat> softmax(Mat&& x) {
  if (x.size() == 0) {
    return {};
  }
  decltype(auto) x_ref = to_ref(std::forward<Mat>(x));
  const auto s = softmax(value_of(x_ref)).eval();
  const auto d_in = x_ref.d().eval();

  return to_fvar(s, s.array() * (d_in.array() - s.cwiseProduct(d_in).sum()));
}
```

File: test/unit/math/fwd/fun/log_softmax_test.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+   * exp(-1000) underflows to zero in double precision. Therefore:
+   *
+   *   log(softmax(x)[2])
+   *
+   * gives -inf while
+   *
+   *   x[2] - log_sum_exp(x)
+   *
+   * correctly retains the log-probability.
+   */
```
Comment (andrjohns) [inline test/unit/math/fwd/fun/log_softmax_test.cpp:30]:
```suggestion
```

Comment isn't really necessary, since the test title already specifies that it's testing a large range

File: test/unit/math/fwd/fun/log_softmax_test.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+  EXPECT_TRUE(std::isfinite(y.val()(0)));
+  EXPECT_TRUE(std::isfinite(y.val()(1)));
+  EXPECT_TRUE(std::isfinite(y.val()(2)));
```
Comment (andrjohns) [inline test/unit/math/fwd/fun/log_softmax_test.cpp:34]:
```suggestion
```

Will be covered by the value tests below

File: test/unit/math/fwd/fun/log_softmax_test.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
+#include <stan/math/fwd.hpp>
+#include <gtest/gtest.h>
+
+#include <cmath>
+
+TEST(AgradFwd, log_softmax_large_dynamic_range) {
```
Comment (andrjohns) [inline test/unit/math/fwd/fun/log_softmax_test.cpp:6]: You can take advantage of some of utility/helper functions for matrix equalities to simplify these tests. For example, this test could be reduced down to:

```cpp
TEST(AgradFwd, log_softmax_large_dynamic_range) {
  using stan::math::fvar;
  using stan::math::vector_fd;
  using stan::math::log_softmax;

  Eigen::VectorXd val(3);
  val << 0, -100, -1000;

  Eigen::VectorXd d(3);
  d << 1, 2, 3;

  vector_fd x = stan::math::to_fvar(val, d);

  auto y = log_softmax(x);

  EXPECT_MATRIX_EQ(y.val(), val);
  EXPECT_MATRIX_EQ(y.d(), d.array() - (1.0 + 2.0 * std::exp(-100.0)));
}
``` 

This also applies to the other tests in the PR, so I won't duplicate the review comments there

File: test/unit/math/fwd/fun/log_sum_exp_test.cpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
+#include <stan/math/fwd.hpp>
+#include <gtest/gtest.h>
+
+#include <cmath>
+
+TEST(AgradFwd, log_sum_exp_derivative_does_not_overflow) {
```
Comment (andrjohns) [inline test/unit/math/fwd/fun/log_sum_exp_test.cpp:6]: You can remove all comments from this one since the test name covers it

**Review by andrjohns — CHANGES_REQUESTED** [review]
Thanks for the contribution! Looking pretty good, just some extra simplifications you can take advantage of and a couple of performance improvements

File: stan/math/fwd/fun/log_softmax.hpp (hunk 2 above)
Comment (helske) [reply stan/math/fwd/fun/log_softmax.hpp:39]: Thanks! Do you think this version of relying on separate `log_sum_exp(x_val)` and `softmax(x_val)` calls is preferable, or should we compute them jointly here? The latter would avoid computing maxCoeff() and exp() twice as we could do

```
template <typename Mat, require_eigen_vt<is_fvar, Mat>* = nullptr>
inline plain_type_t<Mat> log_softmax(Mat&& x) {
  if (x.size() == 0) {
    return {};
  }
  decltype(auto) x_ref = to_ref(std::forward<Mat>(x));
  const auto x_val = value_of(x_ref).eval();
  const auto d_in = x_ref.d().eval();

  const double max_x = x_val.maxCoeff();
  const auto theta = (x_val.array() - max_x).exp().eval();
  const double theta_sum = theta.sum();
  const double lse = max_x + std::log(theta_sum);
  const double d_adj = theta.matrix().cwiseProduct(d_in).sum() / theta_sum;

  return to_fvar(x_val.array() - lse, d_in.array() - d_adj);
```

Currently `fwd::log_sum_exp` and `rev::log_sum_exp` also call both `softmax()` and `log_sum_exp()`, so this touches those ones as well.

Comment (helske) [reply stan/math/fwd/fun/log_softmax.hpp:39]: Not to rush, but just pinging @andrjohns in case there isn’t a notification for these comments otherwise?

Comment (andrjohns) [reply stan/math/fwd/fun/log_softmax.hpp:39]: Apologies for the delay! Yes it would be great to reduce as much duplicate computation as possible, as long as it doesn't end up duplicating too much code. If you find that you're using the same constructions across multiple headers then you can just create a separate helper function.
